pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,2039 @@
+                const std::vector<std::complex<Precision>> &gate_matrix) {
+        std::vector<CFP_t> matrix_cu(gate_matrix.size());
+
+        for (std::size_t i = 0; i < gate_matrix.size(); i++) {
+            matrix_cu[i] =
+                cuUtil::complexToCu<std::complex<Precision>>(gate_matrix[i]);
+        }
+
+        if (gate_matrix.empty()) {
+            std::string message = "Currently unsupported observable";
+            throw LightningException(message.c_str());
+        }
+
+        // Wire order reversed to match expected custatevec wire ordering for
+        // tensor observables.
+        auto &&local_wires = std::vector<size_t>{wires.rbegin(), wires.rend()};
+        PL_CUDA_IS_SUCCESS(cudaDeviceSynchronize());
+        auto expect_val =
+            getExpectationValueDeviceMatrix(matrix_cu.data(), local_wires).x;
+        return expect_val;
+    }
+
+    /**
+     * @brief Get expectation value for a sum of Pauli words. This function
+     * accepts a vector of words, where each word contains a set of Pauli
+     * operations along with their corresponding wires and coefficients of the
+     * Hamiltonian. The expval calculation is performed based on the word and
+     * its corresponding target wires. The distribution of target wires can be
+     * categorized into three different types:1. All target wires are local, and
+     * no MPI operation is required. 2. Some target wires are located at the
+     * global wires, and there are sufficient local wires available for bit swap
+     * operations. 3. Some target wires are located at the global wires, and
+     * there are insufficient local wires available for bit swap operations. For
+     * the first scenario, the `expvalOnPauliBasis` method can be called
+     * directly to obtain the `expval`. In the second scenario, bit swap
+     * operations are necessary before and after the `expvalOnPauliBasis`
+     * operation to get the `expval`. In the third scenario, a temporary state
+     * vector is created to calculate the `bra`, and then the inner product is
+     * used to calculate the `expval`. This function here will check the
+     * corresponding target wires of all word first. If all target wires for
+     * each word are local, the `expvalOnPauliBasis` will be called just once.
+     * Otherwise, each word will be looped and corresponding expval calculation
+     * methods will be adopted.
+     *
+     * @param pauli_words Vector of Pauli-words to evaluate expectation value.
+     * @param tgts Coupled qubit index to apply each Pauli term.
+     * @param coeffs Numpy array buffer of size |pauli_words|
+     * @return auto Expectation value.
+     */
+    auto getExpectationValuePauliWords(
+        const std::vector<std::string> &pauli_words,
+        const std::vector<std::vector<std::size_t>> &tgts,
+        const std::complex<Precision> *coeffs) {
+        std::vector<double> expect_local(pauli_words.size());
```

Comment:
I assume we need this to be double, or should it be using `PrecisionT`?

Reply:
see below